### PR TITLE
[URGENT] Virtualization: code enhancement to make test more robust over ipmi and bug fix

### PR DESCRIPTION
**Pull request by alice-suse**
Changes:
  *) change login_console script to let grub menuentry selection for xen more stable, by sending a key 'down' to grub to stop timer counting down
  *) let system_role test show up for sles12sp2 host installation in sles12sp3+ milestone test
      Currently system_role.pm is not properly loaded for those test installing host to sles12sp2 in sles12sp3 release test, so change code to detect not only those installed to the test release, but also those with INSTALL_TO_OTHERS.

Verification link:
login_console grub enhancement:
http://147.2.212.135/tests/1017
http://147.2.212.135/tests/1016

system_role loading:
http://147.2.212.135/tests/1033
http://147.2.212.135/tests/1032

### CODE REVIEW

File: tests/virt_autotest/login_console.pm [hunk 1]
```diff
@@ -35,9 +35,12 @@ sub login_to_console() {
     assert_screen([qw(grub2 grub1)], 120);
     if (!get_var("reboot_for_upgrade_step")) {
         if (get_var("XEN") || check_var("HOST_HYPERVISOR", "xen")) {
-            send_key_until_needlematch("virttest-bootmenu-xen-kernel", 'down', 10, 1);
-            send_key 'ret';
+            if (!match_has_tag("virttest-bootmenu-xen-kernel")) {
+                send_key 'down';
```
Comment (dzedro) [inline tests/virt_autotest/login_console.pm]: why send_key 'down' when below is send_key_until_needlematch pressing 'down' anyway

When I look on this whole change in login_console.pm I think it is not needed, maybe I'm wrong,
but the code before will do the same as the new code (pressing down until virttest-bootmenu-xen-kernel is found) or do I miss something ?

Comment (sysrich) [inline tests/virt_autotest/login_console.pm]: I agree with @dzedro , these changes do not make any sense to me

The solution to the problem with menu entry selection requires something like line 51-53 in installation/grub_test.pm where we do that to stop the timeout so we have more time for the needles to match

Arbitrarily pressing down is NOT a good solution and will break things when boot menu orders change

File: products/sle/main.pm [hunk 2]
```diff
@@ -469,7 +469,7 @@ sub load_inst_tests() {
             && sle_version_at_least('12-SP2')
             && is_server()
             && (!is_sles4sap() || is_sles4sap_standard())
-            && install_this_version())
+            && (install_this_version() || install_to_other_at_least('12-SP2')))
```
Comment (okurz) [inline products/sle/main.pm]: That does not relate at all to "make test more robust over ipmi" and deserves a comment

File: tests/virt_autotest/login_console.pm (hunk 1 above)
Comment (alice-suse) [inline tests/virt_autotest/login_console.pm]: * Logic is correct and needed, reasons are:
  First, you should see it is for xen only(kvm is the default first menuentry, no send key is needed). And then we check it is not the desired one with needle 'virttest-bootmenu-xen-kernel', so we send key down to stop grub timeout.
  Why not use original send_key_until_needle match is that most times the grub times out before the ipmi machine responds to the send_key down event, so the default first menuentry is booted. This results a lot of failures in our test. This is also why I call it a code enhancement. The original code logic is correct, but it just does not work in the real test over ipmi.

* But I will change code to use send_key 'up' as grub_test.pm do, since it is more simple, and do not change anything in effect except stop grub timeout

File: products/sle/main.pm (hunk 2 above)
Comment (alice-suse) [inline products/sle/main.pm]: Yes, you are right. This is a bug fix, rather than a enhancement to make test moe robust over ipmi. I just combine the two fix together.

I will add comment for it.

**Comment by alice-suse** [comment]
Code rework according to comments is done.
Two changes:
 * add comment for system_role loading condition change
 * simplify enhancement to avoid grub timeout for xen

**Comment by alice-suse** [comment]
@sysrich would you please review again?

File: lib/utils.pm [hunk 3]
```diff
@@ -397,6 +398,23 @@ sub sle_version_at_least {
     die "unsupported SLE $version_variable $version in check";
 }
 
+sub install_to_other_at_least {
```
Comment (okurz) [inline lib/utils.pm]: I am sorry but the name does not make much sense to me. Care to add a (better) comment on top describing what this is supposed to do? Also, since so far this sub is only used in products/sle/main.pm I recommend to define it just there. We can still move it when we want to use it elsewhere

File: products/sle/main.pm (hunk 2 above)
Comment (okurz) [inline products/sle/main.pm]: I'm sorry I still don't understand. What is special about 12-SP2?

File: tests/virt_autotest/login_console.pm [hunk 4]
```diff
@@ -35,6 +35,7 @@ sub login_to_console() {
     assert_screen([qw(grub2 grub1)], 120);
     if (!get_var("reboot_for_upgrade_step")) {
         if (get_var("XEN") || check_var("HOST_HYPERVISOR", "xen")) {
+            send_key 'up';
```
Comment (okurz) [inline tests/virt_autotest/login_console.pm]: please add a comment on top. Also, I am wondering if you are not replicating the logic from e.g. tests/installation/grub_test.pm here too much and should rather use more of the same code

File: lib/utils.pm (hunk 3 above)
Comment (alice-suse) [inline lib/utils.pm]: I will add comment.

But I recommend to keep it in utils. Because after we introduce the parameter "INSTALL_TO_OTHERS", which will force to install the system to the version you give, instead of the current PRODUCT the test is put in. So theoretically, all future conditions which do  or not do something according to the version the system is in should use the REAL version it is , rather than the PRODUCT version the test is put.

File: tests/virt_autotest/login_console.pm (hunk 4 above)
Comment (alice-suse) [inline tests/virt_autotest/login_console.pm]: To the comment for line 472, the system_role.pm should be loaded for installation to sle12sp2 and its above. Without my change, the original code load this file only for test put in sles12sp2/sles12sp3 job group and without INSTALL_TO_OTHERS. This is not correct for our tests that are put in sles12sp3 job group, but install to sles12sp2 at first via INSTALL_TO_OTHERS.

Comment (alice-suse) [inline tests/virt_autotest/login_console.pm]: To the comment for line 38 of send_key up, I will add comment. And of course we do a lot of special handlings regarding different virtualization scenarios in this file comparing with grub_test.pm.

**Comment by alice-suse** [comment]
Code rework is done.

**Comment by okurz** [comment]
I prefer if you remove the "URGENT" tag from the PR subject line and split the commit into (at least) two with a proper description for each. "Make more robust over IPMI" does not really apply for both. I still don't understand your explanations regarding the function in lib/utils.pm but I am fine to keep it there for now

File: products/sle/main.pm (hunk 2 above)
Comment (coolo) [inline products/sle/main.pm]: if it's not special why is listed 3 lines above

Comment (okurz) [inline products/sle/main.pm]: what is listed 3 lines above is also "special" and I don't understand why sle-12-sp2 is mentioned there

Comment (alice-suse) [inline products/sle/main.pm]: Before sle12sp2, there is no system_role selection during installation, so this test module should only be loaded for those sle12sp2+

Comment (okurz) [inline products/sle/main.pm]: yes, so comment this in the code?

**Comment by alice-suse** [comment]
I split it into two commits, one is for code enhancement and the other is for bug fix.

This PR is quite urgent. We are counting on it for the sles12sp3 beta1 testing.  Please help to review again.

File: products/sle/main.pm (hunk 2 above)
Comment (alice-suse) [inline products/sle/main.pm]: done
